ui/jobs/event_emitter.ts

```diff
@@ -0,0 +1,231 @@
+
+    switch (type) {
+      case logDefinitions.GameLog.type:
+        this.emit(
+          'log/game',
+          normalizeLogLine(ev.line, logDefinitions.GameLog.fields),
+          ev.line,
+          ev.rawLine,
+        );
+        break;
+      case logDefinitions.GainsEffect.type: {
+        const matches = normalizeLogLine(ev.line, logDefinitions.GainsEffect.fields);
+        if (matches.effectId)
+          this.emit('effect/gain', matches.effectId, matches);
+        break;
+      }
+      case logDefinitions.LosesEffect.type: {
+        const matches = normalizeLogLine(ev.line, logDefinitions.LosesEffect.fields);
+        if (matches.effectId)
+          this.emit('effect/lose', matches.effectId, matches);
+        break;
+      }
+      case logDefinitions.NetworkDoT.type: {
+        const matches = normalizeLogLine(ev.line, logDefinitions.NetworkDoT.fields);
+        const damage = parseInt(matches.damage ?? '0', 16); // damage is in hex
+        if (matches.which === 'DoT')
+          this.emit('tick/dot', damage, matches);
+        else if (matches.which === 'HoT')
+          this.emit('tick/hot', damage, matches);
+        break;
+      }
+
+      default:
+        break;
+    }
+  }
+
+  processEnmityTargetData({ Target: target }: OverlayEventResponses['EnmityTargetData']): void {
+    if (target) {
+      this.emit('battle/target', {
+        name: target.Name,
+        distance: target.Distance,
+        effectiveDistance: target.EffectiveDistance,
+      });
+    } else {
+      this.emit('battle/target');
+    }
+  }
+}
+
+/**
+ * Track DoTs that was applied to mobs.
+ *
+ * Emit events when DoT ticks on your main target.
+ *
+ * The mechanism of this tracker is to find out the target that player
+ * has applied DoT to and attacked with some actions.
+ * (Which we call it the "last attacked DoTed target", is the boss in most cases.)
+ * For a Bard, if there are a boss and some adds, the Bard would likely
+ * keep the DoT on the boss while attacking the adds without any DoT.
+ *
+ * @example
+ *
+ * const tracker = new DoTTracker({ emitter: emitter, player: player});
+ * tracker.onTick([EffectId.Stormbite, EffectId.CausticBite], (targetId) => {
+ *   // do something like update repertoire timer.
+ * });
+ */
+export class DotTracker extends EventEmitter<{ tick: (targetId?: string) => void }, DotTracker> {
+  ee: JobsEventEmitter;
+  player: Player;
+  trackedDoTs: string[];
+
+  targets: string[];
+  lastAttackedTarget?: string;
+
+  constructor(o: {
+    emitter: JobsEventEmitter;
+    player: Player;
+  }) {
+    super();
+
+    this.ee = o.emitter;
+    this.player = o.player;
+    this.trackedDoTs = [];
+
+    this.targets = [];
+
+    this.registerListeners();
+  }
+
+  private registerListeners(): void {
+    this.player.on('effect/gain', (id, { sourceId, targetId }) => {
+      if (
+        targetId?.startsWith('4') &&
+        parseInt(sourceId ?? '0', 16) === this.player.id &&
```

Comment:
What about storing the hex id of the player instead (or in addition)? I feel like there are a lot of `parseInt(x, 16)` around that could be simpler if we kept the hex id around.

Reply:
In fact, my fealing is to store all id in decimal would be better.
I have already trapped in several times that comparing with hex numbers in string that one is in lower cases and the other one is upper cases so cause bugs, at least all in decimal make it consistent?

Reply:
In the next version of the ffxiv plugin, all ids will be uppercase in all log lines (network and parsed).  So, I think it will be consistent in the future, if that helps reassure you.

It just seems like if logs lines have a hex id, then it's more straightforward to compare a hex id against the log line rather than doing a hex->decimal conversion on every line.  That's maybe premature optimization, though, but I think it also makes the code simpler (and prevents any bugs where somebody forgot to include the `, 16` part of `parseInt`).

Reply:
Wow, that is really good news.
maybe I can call them `id` and a `idHex`(in uppercase), and use the `idHex` for comparing for the future.